shell/platform/android/io/flutter/plugin/editing/TextEditingDelta.java

```diff
@@ -0,0 +1,187 @@
+
+    // Is deleting/inserting at the end of a composing region.
+    final boolean isDeletingInsideComposingRegion = !isReplacedByShorter && start + tbend < end;
+    final boolean isInsertingInsideComposingRegion = start + tbend > end;
+
+    // To consider the cases when autocorrect increases the length of the text being composed by
+    // one, but changes more than one character.
+    final boolean isOriginalComposingRegionTextChanged =
+        (isCalledFromDelete || isDeletingInsideComposingRegion || isReplacedByShorter)
+            || !oldEditable
+                .subSequence(start, end)
+                .toString()
+                .equals(tb.subSequence(tbstart, end - start).toString());
+
+    final boolean isEqual = oldEditable.equals(newEditable);
+
+    // A replacement means the original composing region has changed, anything else will be
+    // considered an insertion.
+    final boolean isReplaced =
+        isOriginalComposingRegionTextChanged
+            && (isReplacedByLonger || isReplacedBySame || isReplacedByShorter);
+
+    if (isEqual) {
+      Log.v(TAG, "A TextEditingDelta for an TextEditingDeltaType.nonTextUpdate has been created.");
+      setDeltas(oldEditable, "", "TextEditingDeltaType.nonTextUpdate", -1, -1);
+    } else if (isCalledFromDelete || isDeletingInsideComposingRegion) {
+      Log.v(TAG, "A TextEditingDelta for a TextEditingDeltaType.deletion has been created.");
+      final int startOfDelete;
+      if (isDeletionGreaterThanOne) {
+        startOfDelete = start;
+      } else {
+        startOfDelete = end - 1;
+      }
+
+      setDeltas(oldEditable, "", "TextEditingDeltaType.deletion", startOfDelete, end);
+    } else if ((start == end || isInsertingInsideComposingRegion)
+        && !isOriginalComposingRegionTextChanged) {
+      Log.v(TAG, "A TextEditingDelta for an TextEditingDeltaType.insertion has been created.");
+      setDeltas(
+          oldEditable,
+          tb.subSequence(end - start, tbend).toString(),
+          "TextEditingDeltaType.insertion",
+          end,
+          end);
+    } else if (isReplaced) {
+      Log.v(TAG, "A TextEditingDelta for a TextEditingDeltaType.replacement has been created.");
+      setDeltas(
+          oldEditable,
+          tb.subSequence(tbstart, tbend).toString(),
+          "TextEditingDeltaType.replacement",
+          start,
+          end);
+    }
+  }
+
+  @VisibleForTesting
+  public TextEditingDelta(
+      CharSequence oldText,
+      CharSequence deltaText,
+      CharSequence deltaType,
+      int deltaStart,
+      int deltaEnd,
+      int selectionStart,
+      int selectionEnd,
+      int composingStart,
+      int composingEnd) {
+    newSelectionStart = selectionStart;
+    newSelectionEnd = selectionEnd;
+    newComposingStart = composingStart;
+    newComposingEnd = composingEnd;
+
+    setDeltas(oldText, deltaText, deltaType, deltaStart, deltaEnd);
+  }
+
+  public CharSequence getOldText() {
```

Comment:
Do you need these getters? The text input plugin only need to be able to build deltas and serialize them to JSON right? Are these for writing tests?

Reply:
Yes there are now purely for tests.

Reply:
Nit: Maybe write a comment or comments stating that these are for testing only.

Reply:
I don't know java too well but do you need expose them as `public` if the test is in the same package?

Reply:
Yeah I think we do. I just tried to switch them to private and it complained.